sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/lang/SqlDelightStructureViewFactory.kt

```diff
@@ -0,0 +1,94 @@
+import com.intellij.navigation.NavigationItem
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.util.Iconable
+import com.intellij.pom.Navigatable
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.squareup.sqldelight.core.SqlDelightTypes
+import com.squareup.sqldelight.core.lang.SqlDelightFile
+import com.squareup.sqldelight.core.lang.util.childrenOfType
+import javax.swing.Icon
+
+class SqlDelightStructureViewFactory : PsiStructureViewFactory {
+  override fun getStructureViewBuilder(psiFile: PsiFile) = object : TreeBasedStructureViewBuilder() {
+    override fun createStructureViewModel(editor: Editor?) = SqlDelightStructureViewModel(psiFile)
+    override fun isRootNodeShown() = false
+  }
+}
+
+internal class SqlDelightStructureViewModel(
+  psiFile: PsiFile
+) : StructureViewModelBase(psiFile, SqlDelightStructureViewElement(psiFile)),
+    StructureViewModel.ElementInfoProvider {
+  override fun isAlwaysLeaf(element: StructureViewTreeElement) = false
+  override fun isAlwaysShowsPlus(element: StructureViewTreeElement) = false
+}
+
+internal class SqlDelightStructureViewElement(
+  private val element: PsiElement
+) : StructureViewTreeElement,
+    ItemPresentation,
+    NavigationItem {
+  override fun getValue() = element
+  override fun getChildren() = when (element) {
+    is SqlDelightFile -> element.childIdentifiers()
+    else -> emptyArray()
+  }
+
+  override fun canNavigateToSource() = (element as Navigatable).canNavigateToSource()
+  override fun canNavigate() = (element as Navigatable).canNavigate()
+  override fun navigate(requestFocus: Boolean) = (element as Navigatable).navigate(requestFocus)
+  override fun getPresentation() = this
+  override fun getName(): String? = when {
+    element is SqlDelightFile -> element.virtualFile?.name
+    element.node.elementType == SqliteTypes.TABLE_NAME -> "CREATE TABLE ${element.text}"
+    element.node.elementType == SqliteTypes.IDENTIFIER -> element.text
+    else -> throw IllegalStateException("Unhandled element $element")
+  }
+
+  override fun getLocationString(): String? = null
+  override fun getIcon(open: Boolean): Icon? = element.getIcon(Iconable.ICON_FLAG_READ_STATUS)
+
+  override fun getPresentableText() = name
+
+  companion object {
+    private fun SqlDelightFile.childIdentifiers(): Array<TreeElement> {
+      val tableNames = childrenOfType(SqlDelightTypes.SQL_STMT_LIST)
+          .flatMap { it.childrenOfType(SqliteTypes.STATEMENT) }
+          .flatMap { it.childrenOfType(SqliteTypes.SQL_STMT) }
+          .flatMap { it.childrenOfType(SqliteTypes.CREATE_TABLE_STMT) }
+          .flatMap { it.childrenOfType(SqliteTypes.TABLE_NAME) }
+      val stmtIdentifiers = childrenOfType(SqlDelightTypes.SQL_STMT_LIST)
+          .flatMap { it.childrenOfType(SqlDelightTypes.STMT_IDENTIFIER) }
+          .flatMap { it.childrenOfType(SqliteTypes.IDENTIFIER) }
+      return (tableNames + stmtIdentifiers)
```

Comment:
I would also include the views, triggers, virtual tables and indexes in the structure view

```
val viewNames = statementList.statementList.flatMap { it.sqlStmt.createView.viewName }
// etc for other 3
```